halo2-base/src/gates/builder/mod.rs

```diff
@@ -492,25 +489,40 @@ pub struct GateCircuitBuilder<F: ScalarField> {
     pub builder: RefCell<GateThreadBuilder<F>>, // `RefCell` is just to trick circuit `synthesize` to take ownership of the inner builder
     /// Break points for threads within the circuit
     pub break_points: RefCell<MultiPhaseThreadBreakPoints>, // `RefCell` allows the circuit to record break points in a keygen call of `synthesize` for use in later witness gen
+    /// Configuration parameters for the circuit shape
+    pub config_params: BaseConfigParams,
 }
 
 impl<F: ScalarField> GateCircuitBuilder<F> {
     /// Creates a new [GateCircuitBuilder] with `use_unknown` of [GateThreadBuilder] set to true.
-    pub fn keygen(builder: GateThreadBuilder<F>) -> Self {
-        Self { builder: RefCell::new(builder.unknown(true)), break_points: RefCell::new(vec![]) }
+    pub fn keygen(builder: GateThreadBuilder<F>, config_params: BaseConfigParams) -> Self {
```

Comment:
A question - why does `GateThreadBuilder` need to be created outside `GateCircuitBuilder`?

Reply:
It could be created along with `GateCircuitBuilder`, but it needs to be in a `RefCell` because in `synthesize` you only get to borrow `&self` and not `&mut self`